cpp/src/arrow/dataset/discovery.h

```diff
@@ -238,5 +238,27 @@ class ARROW_DS_EXPORT FileSystemDatasetFactory : public DatasetFactory {
   util::optional<util::string_view> RemovePartitionBaseDir(util::string_view path);
 };
 
+class ARROW_DS_EXPORT SingleFileDatasetFactory : public DatasetFactory {
```

Comment:
I was aware of the changes in ARROW-8318 but at this time I can remember what stops me from removing SingleFileDataset and migrating to FileSystemDataset. I'll have another try next week.

Reply:
One question: can we add a construction function `FileSystemDatasetFactory::Make(std::string uri, std::shared_ptr<FileFormat> format, FileSystemFactoryOptions options)` to `FileSystemDatasetFactory`, for reading a single file from an uri? This is similar to functionality of `SingleFileDataset`.

The original reason of creating `SingleFileDataset` is to avoid passing file system object via JNI. Passing such objects can make the bridge even complicated.